### PR TITLE
FIT-229 Implement Shared Library for Automated Licensing Phase Checks

### PR DESCRIPTION
**Pull request by SergiiRedchuk**
Fixed commands for generating licenses report for front-end projects.
Fixed the git push command.
Added a test for the LicensingSupportTypeDeterminer, hence made the test for LicensingSupport simpler.

### CODE REVIEW

File: src/gov/ca/cwds/jenkins/licensing/LicensingSupportTypeDeterminer.groovy [hunk 1]
```diff
@@ -10,19 +9,20 @@ class LicensingSupportTypeDeterminer {
     this.pipeline = pipeline
   }
 
-  LicensingSupportType determineLicensingSupportType() {
-    def projectTypesDeterminer = new ProjectTypesDeterminer(pipeline)
-    def projectTypes = projectTypesDeterminer.determineProjectTypes('.')
+  LicensingSupportType determineLicensingSupportType(buildMetadata) {
+    def licensingSupportType = LicensingSupportType.NONE
+    def projectTypes = buildMetadata.projectTypes()
```
Comment (edgibbs) [inline src/gov/ca/cwds/jenkins/licensing/LicensingSupportTypeDeterminer.groovy]: I think this is cleaner with just calling ProjectTypesDeterminer, if you need the project path just use `pipeline.env.WORKSPACE`

File: test/gov/ca/cwds/jenkins/licensing/LicensingSupportTypeDeterminerSpecification.groovy [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  class PipeLineScript {
+    PipeLineScript() {
+    }
+
+    def sh(hash) {}
+
+    def echo(String) {}
+  }
+
+  def "When a Java project without the Hierynomus License gradle plugin then the 'NONE' Licensing Support Type is determined"() {
```
Comment (edgibbs) [inline test/gov/ca/cwds/jenkins/licensing/LicensingSupportTypeDeterminerSpecification.groovy]: I like the use of the optional arguments to describe the given/when/then context.  No need to change the spec descriptions, but you could probably cut it back by moving a lot of the descriptions to the given/when/then descriptions something like:

```groovy
def "with a java project and no plugin"() {
given: "a pipeline"
when: "determining support"
then: "the 'None' Licensing support Type is determined"
}
```

File: src/gov/ca/cwds/jenkins/licensing/LicensingSupport.groovy [hunk 3]
```diff
@@ -8,32 +8,29 @@ class LicensingSupport {
...
-  LicensingSupport(pipeline) {
+  LicensingSupport(pipeline, runtimeGradle = null, dockerImage = null) {
     this.pipeline = pipeline
+    this.runtimeGradle = runtimeGradle
+    this.dockerImage = dockerImage
+    this.licensingSupportTypeDeterminer = new LicensingSupportTypeDeterminer(pipeline)
   }
 
-  def updateLicenseReport(branchName, sshCredentialsId, runtimeGradle = null) {
+  def updateLicenseReport(branchName, sshCredentialsId, buildMetadata) {
```
[the remark this answers is not in the document]
Comment (SergiiRedchuk) [reply src/gov/ca/cwds/jenkins/licensing/LicensingSupport.groovy]: I did not create BuildMetadata and I'm reusing it to get projectTypes(). Refactoring the class is out of current story scope.

File: src/gov/ca/cwds/jenkins/licensing/LicensingSupport.groovy [hunk 4]
```diff
@@ -8,32 +8,29 @@ class LicensingSupport {
...
-    pipeline.echo("Detected Licensing Support Type: ${licensingSupportType.title}")
-    if (LicensingSupportType.NONE == licensingSupportType) {
-      throw new Exception(MSG_NO_LICENSING_SUPPORT)
-    }
-  }
-
-  private def generateLicenseReport(runtimeGradle = null) {
+  private def generateLicenseReport(buildMetadata) {
     pipeline.echo 'Generating License Information'
+    def licensingSupportType = licensingSupportTypeDeterminer.determineLicensingSupportType(buildMetadata)
```
[the remark this answers is not in the document]
Comment (SergiiRedchuk) [reply src/gov/ca/cwds/jenkins/licensing/LicensingSupport.groovy]: Please elaborate.
Any proposed lines of code perhaps?

File: test/gov/ca/cwds/jenkins/licensing/LicensingSupportTypeDeterminerSpecification.groovy (hunk 2 above)
Comment (SergiiRedchuk) [reply test/gov/ca/cwds/jenkins/licensing/LicensingSupportTypeDeterminerSpecification.groovy]: Do you mean, to make the method names shorter?

File: src/gov/ca/cwds/jenkins/licensing/LicensingSupport.groovy [hunk 5]
```diff
@@ -8,32 +8,29 @@ class LicensingSupport {
   final def MSG_NO_LICENSING_SUPPORT = 'No known Licensing Support is found in the project'
 
   def pipeline
-  def licensingSupportType
+  def runtimeGradle
+  def dockerImage
+  def licensingSupportTypeDeterminer
 
-  LicensingSupport(pipeline) {
+  LicensingSupport(pipeline, runtimeGradle = null, dockerImage = null) {
```
[the remark this answers is not in the document]
Comment (SergiiRedchuk) [reply src/gov/ca/cwds/jenkins/licensing/LicensingSupport.groovy]: Currently agreed to use the map as the least evil thing.

File: test/gov/ca/cwds/jenkins/licensing/LicensingSupportTypeDeterminerSpecification.groovy (hunk 2 above)
Comment (edgibbs) [reply test/gov/ca/cwds/jenkins/licensing/LicensingSupportTypeDeterminerSpecification.groovy]: Maybe just a little less verbose, but they're still good.

File: src/gov/ca/cwds/jenkins/licensing/LicensingSupport.groovy [hunk 6]
```diff
@@ -43,27 +40,35 @@ class LicensingSupport {
...
+      runGitSshCommand('git commit -m "updated license info"')
+      runGitSshCommand('git push origin HEAD:master', true)
     }
   }
 
-  private def runSshCommand(command, failOnNonZeroStatus = false) {
+  private def runGitSshCommand(command, failOnNonZeroStatus = false) {
     // Used to avoid known_hosts addition, which would require each machine to have GitHub added in advance (maybe should do?)
     def cmd = 'GIT_SSH_COMMAND="ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no" ' + command
     def status = pipeline.sh script: cmd, returnStatus: true
```
[the remark this answers is not in the document]
Comment (SergiiRedchuk) [reply src/gov/ca/cwds/jenkins/licensing/LicensingSupport.groovy]: Removed the parameter and explicit checking of return status. Testing...

**Pull request merged**